### PR TITLE
Fix array oob crash (IOS-257)

### PR DESCRIPTION
**Pull request by kimar**
Supposed to fix an out-of-bounds crash:

Bildschirmfoto 2024-04-23 um 14 26 12

### CODE REVIEW

File: Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift [hunk 1]
```diff
@@ -105,9 +105,10 @@ extension HomeTimelineViewModel.LoadLatestState {
 
         guard let viewModel else { return }
 
-        let latestFeedRecords = viewModel.dataController.records.prefix(APIService.onceRequestStatusMaxCount)
 
-        Task {
+        Task { @MainActor in
+            let latestFeedRecords = viewModel.dataController.records.prefix(APIService.onceRequestStatusMaxCount)
```
Comment (zeitschlag) [inline Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift:110]: I don't think this has something to do with the index-fix, but still it crossed my mind:

Do you know or remember the reason for `prefix(100)`? A few lines below we use the IDs of `latestFeedRecords` to determine new statuses: A status is new, if its ID isn't among those of max. 100 `latestFeedRecords` (from my understanding). But what if its the 101. or 102.?

I didn't find anything in the git history.

File: Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift [hunk 2]
```diff
@@ -146,7 +147,7 @@ extension HomeTimelineViewModel.LoadLatestState {
                         for (i, record) in newRecords.enumerated() {
                             if let index = oldRecords.firstIndex(where: { $0.status?.reblog?.id == record.id || $0.status?.id == record.id }) {
                                 oldRecords[index] = record
-                                if newRecords.count > index {
+                                if newRecords.count > i {
```
Comment (zeitschlag) [inline Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift:150]: Great catch!

**Review by zeitschlag — APPROVED** [review]
As I said: I could reproduce this bug when I lowered the `onceRequestStatusMaxCount` to `10` and this PR (`i` instead of `index`) fixed it. Thank you for that!

File: Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift (hunk 1 above)
Comment (zeitschlag) [reply Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift:110]: I'm asking as using all the records (and not only the `n` most recent onces) seems to fix the weird timeline-sorting issue I mentioned ... somewhere (although some posts are missing then in between. This is super strange)

Comment (DK1DA) [reply Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift:110]: Maybe the hard-coded value is related to the max result count of the (paginated) API? There should be a "Load missing posts" divider if there is a gap in between.

Comment (zeitschlag) [reply Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift:110]: Yes, I've seen that cell before, but it doesn't show up.

`limit` in the API-docs defaults to 20 and has a maximum value of 40. So using 100 as clientside default doesn't make a lot of sense to me.

I'll look into it :)

**Pull request merged**